src/peakWatcher.cpp

```diff
@@ -717,6 +730,172 @@ void report(int watcherIndex, int channel) {
+  WATCHER TRACKFX {someGuid} 5 TYPE 0 FOLLOW 0 LEVEL -5.0 HOLD -1 NOTIFY 1 1
+>
+ */
+
+const char CONFIG_HEADER[] = "<OSARA_PEAKWATCHER";
+const char CONFIG_FOOTER[] = ">";
+
+bool processExtensionLine(const char* line, ProjectStateContext* ctx,
+	bool isUndo, struct project_config_extension_t* reg
+) {
+	if (isUndo || strcmp(line, CONFIG_HEADER) != 0) {
+		return false;
+	}
+	stop();
+	ReaProject* project = GetCurrentProjectInLoadSave();
+	for (int w = 0; ; ++w) {
+		char data[500];
+		ctx->GetLine(data, sizeof(data));
+		if (strcmp(data, CONFIG_FOOTER) == 0) {
+			break;
+		}
+		if (w >= NUM_WATCHERS) {
+			continue;
+		}
+		istringstream input(data);
+		string word;
+		input >> word;
+		if (word != "WATCHER") {
+			continue;
+		}
+		Watcher& watcher = watchers[project][w];
+		watcher.reset();
+		input >> word;
+		if (word == "NONE") {
+			watcher.target = NoTarget();
+		} else if (word == "TRACK") {
+			input >> word;
+			if (MediaTrack* track = getTrackFromGuidStr(project, word)) {
+				watcher.target = track;
+			}
+		} else if (word == "TRACKFX") {
+			input >> word;
+			int fx;
+			input >> fx;
+			if (MediaTrack* track = getTrackFromGuidStr(project, word)) {
+				watcher.target = TrackFx(track, fx);
+			}
+		}
+		input >> word;
+		if (word == "TYPE") {
+			input >> watcher.levelType;
+		}
+		input >> word;
+		if (word == "FOLLOW") {
+			input >> word;
+			watcher.follow = word == "1";
+		}
+		input >> word;
+		if (word == "LEVEL") {
+			input >> watcher.notifyLevel;
+		}
+		input >> word;
+		if (word == "HOLD") {
+			input >> watcher.hold;
+		}
+		input >> word;
+		if (word == "NOTIFY") {
+			for (auto& channel : watcher.channels) {
+				input >> word;
+				channel.notify = word == "1";
+			}
+		}
+	}
+	if (isWatchingAnything()) {
+		start();
+	}
+	return true;
+}
+
+void saveExtensionConfig(ProjectStateContext* ctx, bool isUndo,
+	struct project_config_extension_t* reg
+) {
+	if (isUndo) {
+		return;
+	}
+	ctx->AddLine(CONFIG_HEADER);
+	ReaProject* project = GetCurrentProjectInLoadSave();
+	for (Watcher& watcher : watchers[project]) {
+		ostringstream out;
+		out << "WATCHER";
+		if (holds_alternative<NoTarget>(watcher.target)) {
+			out << " NONE";
+		} else if (MediaTrack** track = get_if<MediaTrack*>(&watcher.target)) {
+			out << " TRACK " << getTrackGuidStr(project, *track);
+		} else if (TrackFx* tfx = get_if<TrackFx>(&watcher.target)) {
+			out << " TRACKFX " << getTrackGuidStr(project, tfx->first) <<
+				" " << tfx->second;
+		}
+		out << " TYPE " << watcher.levelType;
+		out << " FOLLOW " << (int)watcher.follow;
+		out << " LEVEL " << watcher.notifyLevel;
+		out << " HOLD " << watcher.hold;
+		out << " NOTIFY";
+		for (auto& channel : watcher.channels) {
+			out << " " << (int)channel.notify;
+		}
+		ctx->AddLine("%s", out.str().c_str());
+	}
+	ctx->AddLine(CONFIG_FOOTER);
+}
+
+void BeginLoadProjectState (bool isUndo, struct project_config_extension_t* reg){
+	//clean up configuration data for dead projects
```

Comment:
Can we use ValidatePtr instead? That should simplify this a lot, since we could just walk through watchers erasing any associated with invalid projects.

Reply:
Of course. I forgot about that function.

Reply:
I also realised the code was broken because after map::erase the iterator is invalid.

Reply:
Yeah, that map erase behaviour is confusing. In case you aren't aware (sorry if you already are), when you erase, don't increment the iterator, but when you don't erase, do increment the iterator:
```
if (...) {
  map.erase(it);
} else {
  ++it;
}
```
It'd be nice if there were an erase_if method that takes a lambda, but that doesn't exist yet. :(

Reply:
Oh wait, there is [std::erase_if for maps in C++ 20](https://en.cppreference.com/w/cpp/container/map/erase_if). Not sure if our Windows and Mac C++ libraries have it or not.

Reply:
I didn't know about that one, even though it was mentioned further down the stackoverflow page I was reading. :)

Reply:
Heh, I saw an answer on StackOverflow too, but it suggested erase_if didn't exist yet. I guess that StackOverflow post was before C++ 20.